### ISSUE TITLE
UnhandledPromiseRejectionWarning: Error: This job has been dropped by Bottleneck

### ISSUE THREAD

**Issue by wertlex** [SGrondin/bottleneck#136]
Guys, looks like I'm stuck. 
Here is sample which leads to `UnhandledPromiseRejectionWarning` when `highWater` starts triggering strategy and it drops oldest queued task, and I can't figure out if the problem in my code or in bottleneck's.

Following sample:
```javascript
const Bottleneck = require('bottleneck');

async function main() {
  const limiter = new Bottleneck({
    maxConcurrent: 1, // only one full global recalculation at the time is make sense
    strategy: Bottleneck.strategy.LEAK,
    highWater: 1
  });

  const sleep = timeout => {
    return new Promise(resolve => setTimeout(resolve, timeout));
  };

  const onSuccess = () => console.log('ok!');
  const onErr = err => console.log(`err: ${err.message}`);

  const task1Promise = limiter.schedule({ id: 1 }, sleep, 1000);
  const task2Promise = limiter.schedule({ id: 2 }, sleep, 1000);
  const task3Promise = limiter.schedule({ id: 3 }, sleep, 1000);

  await task1Promise.then(onSuccess, onErr);
  await task2Promise.then(onSuccess, onErr);
  await task3Promise.then(onSuccess, onErr);
}

main();
```

Ends up with this error:
```javascript
(node:14388) UnhandledPromiseRejectionWarning: Error: This job has been dropped by Bottleneck
    at Job.doDrop (/Users/wert/Projects/combinvest/awesome-scripts/node_modules/bottleneck/lib/Job.js:57:46)
    at /Users/wert/Projects/combinvest/awesome-scripts/node_modules/bottleneck/lib/Bottleneck.js:397:21
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (/Users/wert/Projects/combinvest/awesome-scripts/node_modules/bottleneck/lib/Bottleneck.js:15:103)
    at _next (/Users/wert/Projects/combinvest/awesome-scripts/node_modules/bottleneck/lib/Bottleneck.js:17:194)
(node:14388) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:14388) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

Am I doing something wrong?

**Comment by SGrondin** [comment]
Yup, so what's happening is:
- your settings mean that Bottleneck will drop the oldest task in the queue if a new task would make the queue be longer than 1 item
- Bottleneck drops that task by rejecting the promise with a BottleneckError exception
- NodeJS starts looking for a `.catch()` on your promise
  - if your promise was created inside of an async function (in your case it is!), NodeJS will also look for a `try/catch`
- NodeJS finds nothing to catch the exception
- your NodeJS app crashes

You need to choose between "Promise style" **or** "async/await style", don't use both styles within the same function. If you put the `async` keyword on a function, then use the async/await style within that function.

async/await style is easier to read, easier to write and it's harder to write bugs with it. Promise style is only better if you need to start multiple promises in parallel (like your example!).

I fixed your code by converting it to one single style (Promise style).
- Removed the `await` keyword because you're already using `.then()` to compose promises.
- Removed the second argument to `.then()` because `.then()` only takes one argument.
- Added `.catch(onErr)` on the promises that can fail.
- Removed the unnecessary `async` keyword on your main function
```js
const Bottleneck = require('bottleneck');

function main() {
  const limiter = new Bottleneck({
    maxConcurrent: 1, // only one full global recalculation at the time is make sense
    strategy: Bottleneck.strategy.LEAK,
    highWater: 1
  });

  const sleep = timeout => {
    return new Promise(resolve => setTimeout(resolve, timeout));
  };

  const onSuccess = () => console.log('ok!');
  const onErr = err => console.log(`err: ${err.message}`);

  const task1Promise = limiter.schedule({ id: 1 }, sleep, 1000);
  const task2Promise = limiter.schedule({ id: 2 }, sleep, 1000);
  const task3Promise = limiter.schedule({ id: 3 }, sleep, 1000);

  task1Promise.then(onSuccess).catch(onErr);
  task2Promise.then(onSuccess).catch(onErr);
  task3Promise.then(onSuccess).catch(onErr);
}

main()
```
Output:
```
err: This job has been dropped by Bottleneck
ok!
ok!
```
If you want 2 jobs to be dropped, then set `highWater` to 0. With `highWater` set to 1, `task1Promise` is immediately started, `task2Promise` goes into the queue and `task3Promise` is the one that's dropped.

In Promise style, if you want to await multiple promises before continuing, use `Promise.all()`. Example:
```js
  const task1Promise = limiter.schedule({ id: 1 }, sleep, 1000).then(onSuccess).catch(onErr);
  const task2Promise = limiter.schedule({ id: 2 }, sleep, 1000).then(onSuccess).catch(onErr);
  const task3Promise = limiter.schedule({ id: 3 }, sleep, 1000).then(onSuccess).catch(onErr);

  return Promise.all([task1Promise, task2Promise, task3Promise])
  .then(() => console.log('All 3 promises have finished!'));
```
Output:
```
err: This job has been dropped by Bottleneck
ok!
ok!
All 3 promises have finished!
```

I hope this helps. I encourage you to practice your asynchronous programming skills by converting Promise code to async/await code and vice versa. There are tons and tons of blog posts that can also help with this.

Good luck!

**Comment by wertlex** [comment]
Well, I've found this question almost 2 years after it was asked, and there is an answer for this.

The thing is what @SGrondin suggest (besides code styling) just hides problem, but not resolves the root cause.
And the root cause is a bit trickier thing than it appears at glance.

Let's consider following part of the original example:
```javascript
  const task1Promise = limiter.schedule({ id: 1 }, sleep, 1000); // (1)
  const task2Promise = limiter.schedule({ id: 2 }, sleep, 1000); // (2)
  const task3Promise = limiter.schedule({ id: 3 }, sleep, 1000); // (3)

  await task1Promise.then(onSuccess, onErr); // (4)
  await task2Promise.then(onSuccess, onErr); // (5)
  await task3Promise.then(onSuccess, onErr); // (6)
```

I will explain one by one what is happening here step by step:
- on the lines 1-3 three promises are created one by one
- job under `task1Promise` will be almost immediately scheduled for the execution
- job under `task2Promise` will put into the queue
- job under `task3Promise` will put into the queue, but also this will evict job under `task2Promise` from the queue. Eventually it will lead to the `task2Promise` rejection
- so, before executing line 4 all the promises are in state `pending`
- on the line 4 few things are happening:
   - `task1Promise` receives `onSuccess` and `onErr` handlers
   -  execution of current function is paused till the time `task1Promise` with all it's handler will be resolved. It will take 1000ms
- however, pausing current function execution doesn't mean whole node process is blocked (after all this is why all this async things was designed for). And since `task2Promise` was evicted from the queue by adding job under `task3Promise`, `task2Promise` will be rejected. And this rejection is happening before line 5 is executing.
- _it mean that while current function is paused on line 4, somewhere in "background" `task2Promise` is rejected. And since line 5 wasn't executed at this time, `task2Promise` don't have rejection handler. And this leads to printing warning._ 

So, all the rejection handlers have to be set _before_ promise is really rejected.

I hope this could be interested for somebody.
